Add delayed on-off manager

nordic-krch opened the pull request:
Add module which allows scheduled request to enable a resource. Module is an extension to an on-off manager.

Compared to an on-off manager request API is extended by `k_timeout_t` parameter:
```
int onoff_delayed_request(struct onoff_delayed_manager *mgr,  struct onoff_delayed_client *cli, k_timeout_t delay)
```
Delayed client is based on on-off client extended by `k_timeout_t` field.

Delayed on-off manager gets startup offset during init and triggers scheduled start earlier than requested to ensure that service is in on state at specified time.

nordic-krch committed:
lib: os: onoff: Extend with onoff_delayed API

Added API  which allows scheduled request to enable a resource.
API is implemented as an extension to an on-off manager.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: Add test for onoff_delayed module

Added suite for new module onoff_delayed. Module is based on
onoff manager.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
There was a potential race condition when taking client from `future_clients` list and calling `onoff_request`. In order to safe switch from pending to active request I had to merge onoff_delayed into onoff manager.

pabigot commented:
Why can't delayed changes to an on-off service state be initiated by either a k_timer or a delayed work item?

nordic-krch commented:
> Why can't delayed changes to an on-off service state be initiated by either a k_timer or a delayed work item?

There are few things:
- only one `_timeout` instance used per manager so less ram used compared to timer, work item per client.
- client is not aware of startup time so does not know when to schedule to have service at requested time
- less code on client side, clean API similar to immediate requests

pabigot commented on lib/os/onoff.c:
```diff
@@ -650,3 +661,185 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
+
+static void timeout_handler(struct _timeout *t);
+
+static struct onoff_delayed_client *get_delayed_cli(sys_snode_t *node)
```
make this `inline`

pabigot commented on lib/os/onoff.c:
```diff
@@ -650,3 +661,185 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
+	return false;
+}
+
+sys_snode_t *sortlist_head_peek(sys_slist_t *list)
```
All sortlist functions should be `static`, and most should be `inline`.

pabigot commented on lib/os/onoff.c:
```diff
@@ -650,3 +661,185 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
+	SYS_SLIST_FOR_EACH_NODE(list, n) {
+		cli = get_delayed_cli(n);
+		/* ticks are negative (abs) so inverted comparison */
+		if (newcli->timeout.ticks > cli->timeout.ticks) {
```
Comparison of absolute timeouts needs to be wrapped in API that abstracts from the hack of using a sign bit to identify an absolute timeout.

pabigot commented on doc/reference/resource_management/index.rst:
```diff
@@ -84,5 +84,20 @@ state.
+**********************
+
+A delayed on-off manager is an extension to :ref:`resource_mgmt_onoff` which
+allows scheduled requests to the service. Manager ensures that service will be
```
"ensures" is a bit optimistic.  All this can provide is a best effort, and if the user is unaware of the delay requirements transitions may always occur too late.

pabigot commented on lib/os/onoff.c:
```diff
@@ -415,19 +417,14 @@ static void process_event(struct onoff_manager *mgr,
 
-int onoff_request(struct onoff_manager *mgr,
-		  struct onoff_client *cli)
+static int request(struct onoff_manager *mgr,
```
I'd like this to be named `request_and_unlock`, but at minimum this should document that the lock will be released when the function returns.

pabigot commented on lib/os/onoff.c:
```diff
@@ -650,3 +661,185 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
+		return 0;
+	}
+
+	k_spin_unlock(&mgr->mgr.lock, key);
```
This function needs a test that covers this path.

pabigot commented on include/sys/onoff_delayed.h:
```diff
@@ -0,0 +1,185 @@
+ * @param cli a pointer to the same client state that was provided
+ * when the operation to be cancelled was issued.
+ *
+ * @retval 0 if successfully cancelled before request delay has past.
```
s/past/passed/

The text about "on successful cancellation" should probably be lifted up into the function description, since there are two distinct return value descriptions both of which have that effect.

pabigot reviewed (commented):
A few comments.  I'm skeptical of the value of this enhancement relative to other potential solutions.

The use case must be a situation where a client is aware of a future need for a resource, and wishes it to be available at a particular time without incurring startup delays.  However, the solution requires that this capability be implemented by the resource manager: it doesn't help somebody using a normal on-off service, or a generic service like a regulator where the manager is internal and you don't know whether there's a delay.

How often does this case occur?  If it's common, wouldn't it be better to have it available for anything that might be delayed?  Perhaps the functionality should be integrated directly into the core on-off manager.

The way `_timeout` is used supports exposing it as an lower-level public API as [suggested](https://github.com/zephyrproject-rtos/zephyr/pull/28891#pullrequestreview-501573862) by @andrewboie.  Certainly maintaining a list of absolute timeouts is a generic capability that should not be implemented within this module.

> client is not aware of startup time so does not know when to schedule to have service at requested time

This is the only functionality gap in the existing solution ([the others](https://github.com/zephyrproject-rtos/zephyr/pull/29584#issuecomment-718424510) are optimizations).  I think that could be addressed by making this information available to the client.  I'd be more comfortable if more complex solutions were delayed until we understand more about how power management will deal with this problem (which may very well not be by using the on-off manager infrastructure).

nordic-krch commented:
@pabigot I've addressed your comments. Additionally, I've added `stop time` which previously was not covered. Stop time covers for asynchronous stop operation but also covers for case when restarting service is costly. It is implemented in the following way: service state is monitored, when switching to on state, first pending scheduled request is scheduled `stop_time` earlier. When service is turning into off state, first pending request is scheduled back to its original time.

Regarding the need for this module: `stop time` makes it impossible to do the same outside of on-off manager. Whether it should be core onoff manager functionality, i don't know. I adds RAM.

carlescufi commented:
@andyross and @andrewboie this is using internal timeout APIs for good reason, but in order to make progress we need your input. If we are to avoid using those, we need an alternative approach that can be used in PRs like this one.

pabigot commented:
As I understand it the requirement is to try to ensure that at absolute time `T` the service will be in state `S` (ON or OFF).  Let's call that a "scheduled state".

The service is aware of two durations: `D_on` is the time required to go from OFF to ON, and `D_off` is the time required to go from ON to OFF.

To meet the user requirement a transition may need to be initiated in advance of the target time `T`.  For example, an on-to-off transition must be initiated at `T-D_on` if coming from an OFF state.

I see two problems with this implementation.  First, the on-off service itself is only aware of active clients: ones that have initiated a request or a release.  Second, the transition deadline is adjusted when the schedule request is made.  The combination of these increase the likelihood that a deadline will not be met unless it's guaranteed that only one client will be active at any time.

Consider a service where `D_on` is 1 s and `D_off` is 5 s, and a client that wants the device to be on at absolute time `T` at 10 s after boot.

As implemented I believe this would record the need to initiate a transition to on at absolute time 9 s.

As long as the service is in state OFF at 9 s this will work as intended.

If something else is using the service, particularly with the `K_NO_WAIT` feature to bypass scheduling entirely, the deadline might fail badly.  For example if a client directly requests the service at 7 s, it will turn on at 8 s.  The client does what it needs, and at 8.5 s initiates a transition to off.  At 9 s the delayed infrastructure submits a request to turn on.  But because the service is being shut off, the scheduled request won't get processed until the transition to off completes at 13.5 s (8.5 s + `D_off`) and it won't complete until 14.5 s (13.5 s + `D_on`), long after the deadline.

(Alternatively, if the direct client doesn't initiate a transition to off until 9.5 s, the scheduled client will have received notification that the service is on at 9 s, earlier than the deadline.  Meaning the client must detect that it's too soon, and can't rely on the notification as a signal that the action that requires the service can be initiated on notification.)

What the infrastructure should do to handle this correctly is record the pending request that the service be on at 10 s, and at 8.5 s when the OFF request is initiated detect that it's supposed to be on at 10 s, and so *not* initiate a transition to OFF because doing so would prevent meeting the deadline.  It should instead update the alarm to fire on behalf of the scheduled request at 10 s, and initiate the transition to ON then.  The transition will be trivial and synchronous, and the client will receive notification of the available service at 10 s as expected.

Doing this would require distinguishing active from pending clients in the on-off service, allowing pending clients to inhibit transitions that would cause them to miss their deadline.  Any change to either active or pending clients would require re-evaluating whether a change to a scheduled transition is needed in addition to the current assessment of whether an immediate transition is needed.

Getting the service state machine right for active clients alone was complex; extending it to support pending clients is more complex.

Since the role of this implementation is to be "best effort" I think it would be much simpler, and no less capable, to add API to the base on-off service like:

    k_timeout_t onoff_delay_to_state(struct onoff_manager *mgr,
                                     bool on, k_timeout_t delay);

with the behavior that it returns a timeout that if used with `k_timer` to initiate a transition to the desired state will complete by the provided `delay` as long as the service is in the opposite state at the time the timeout is reached.  (The timeout should be `K_NO_WAIT` or an absolute timeout determined from `delay` and the time at which the service does the calculation.)

The implementation function should be provided by the transition structure, with the wrapper function simply returning `delay` unchanged in the case where the service doesn't provide a `delay_to_state` function.

I should point out (@ceolin) that the logic described above about taking into account future resource requirements is exactly something that the power management infrastructure is going to need when determine whether it's acceptable to change to another power state.  So if that is implemented in a generic way it may supersede not only this effort but the on-off service as a whole.

nordic-krch commented:
@pabigot sorry, missed your response.

> If something else is using the service, particularly with the K_NO_WAIT feature to bypass scheduling entirely, the deadline might fail badly. For example if a client directly requests the service at 7 s, it will turn on at 8 s. The client does what it needs, and at 8.5 s initiates a transition to off. At 9 s the delayed infrastructure submits a request to turn on. But because the service is being shut off, the scheduled request won't get processed until the transition to off completes at 13.5 s (8.5 s + D_off) and it won't complete until 14.5 s (13.5 s + D_on), long after the deadline.

My approach to that problem was following: added onoff monitor. In the monitor if state changes to `on` first pending request is rescheduled to happen `D_off` earlier than requested. If state changes to `off` pending request is rescheduled to its original time. In your example request is rescheduled to (4s) when service is started. At 4s it will request the service (which is still active) and notify the user. On timeout list is checked for another pending request and based on current state target time is set or target - D_off.

The only issue is that user notification for requested time may happen earlier than requested but that should be ok as user request for no later than. I can add explicit note in `onoff_delayed_request`.

Regarding use of `struct _timeout`: it can be easily replaced with single k_timer. I used `_timeout` because i'm targeting device with limited RAM and `_struct timeout` was meeting my timer requirements.

pabigot commented on doc/reference/resource_management/index.rst:
```diff
@@ -84,5 +84,23 @@ state.
+allows scheduled requests to the service. Manager takes the best effort to
+ensure that service will be in the on state at requested absolute or relative
+time. It takes into account service transision times. Start time should be a
+sum of maximum serivce startup time and maximum interrupt latency of the system
```
I don't see what interrupt latency has to do with this.  The service should provide the maximum time required to initiate and complete a transition, assuming nothing delays it.  The application should be responsible for adjusting that based on its perception of any relevant delays that are  outside the control of the on-off service, whether they're due to interrupts or timeslicing or priority threads or whatever.

I think what you wrote here can be interpreted that way, but this PR seems to lack the extension to the on-off service so the maximum delays can be provided as input to that calculation.  With that, this could be made more clear, e.g. "the value of `delay_on` returned by `onoff_service_get_delays(mgr, &delay_on, &delay_off)` plus any notification delay resulting from other system activities".

Though I suspect most on-off service transitions are asynchronous, so the completion won't be affected by other things going on, only the notification of completion might be delayed and probably not by very much.

pabigot commented on lib/os/onoff.c:
```diff
@@ -650,3 +661,255 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
+ * @retval true if @p t0 is equal or later than @t1
+ * @retval false if @p t1 is later than @p t0
+ */
+static inline bool abs_tick_compare(int64_t t0, int64_t t1)
```
What I meant in my original comment was that if we're going to have API to compare timeouts it belongs at the same level as `K_TIMEOUT_EQ`.  So perhaps `K_TIMEOUT_REL_LT(a,b)` and `K_TIMEOUT_ABS_LT(a,b)`, since comparing a relative and absolute timeout is not trivial. 

And it should operate on timeouts, not `int64_t` (which is not necessarily `k_ticks_t`).

pabigot commented on lib/os/onoff.c:
```diff
@@ -650,3 +661,255 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
+		sys_snode_t *head = sortlist_head_peek(&dmgr->future_clients);
+
+		if (head) {
+			uint32_t offset = state == ONOFF_STATE_TO_ON ?
```
OK, I see how the onoff monitor is used for one of the conditions I identified.

What a monitor-based solution doesn't do is handle the case I identified where to meet the deadline a transition in the on-off service must be inhibited.

Also I'm not entirely convinced that looking only at the head of the clients queue when sorted by adjusted deadline is sufficient, if the start and stop delays are different.  That's partly why I suggest storing the actual deadline: it makes it much easier to see when a state has to be reached, not when something has to be done to reach a state at a deadline that's no longer stored.

And this still seems to care only about transitions to on, while transitions to off are equally important for some use cases.

pabigot reviewed (commented):
OK, I see how the monitor is used.  I'm still not convinced that this solves enough of the problem to be worth doing, and I think it'll become obsolete if the full problem is solved for power management in a way that's reusable, but I won't block this based solely on not liking it.

It will need at least two things that I would probably block for, though.  First, add the missing API to the on-off manager to fetch the expected start and stop transition delays, as these are needed to figure out what to pass to the delayed manager initialization routine.  Based on other time-related things going on I think the durations would ideally support nanosecond resolution and not be dependent on tick rate.  `struct timespec` or `uint64_t` are candidates.  That can be a separate commit and can go in independently of this, as it's a useful extension and probably should have always been provided.  (It may be complex if we have to support services that can't provide a bounded maximum for transition completion---which I'm thinking we should.  Then a negative value in `int64_t` nanoseconds might indicate that condition.)

Then there's this:
> The only issue is that user notification for requested time may happen earlier than requested but that should be ok as user request for no later than.

I believe that puts a large burden on the caller to delay if the action that requires the resource if the action is not safe to initiate early (like a timed transmission or sensor sample): the client would still has to maintain a separate timer infrastructure.  It's much cleaner if the notification comes precisely at the deadline.
